### PR TITLE
URL handling

### LINKED ISSUE #5239: Create installation packages for Enso
This task is automatically imported from the old Task Issue Board and it was originally created by Dmitry Bushev.
Original issue is [here](https://www.pivotaltracker.com/story/show/184153161).
<hr />

## Notes
In order to handle custom `enso://` redirect links required for authentication, we need to install .desktop files in appropriate locations. 
Packaging is required for Windows and Linux platforms. The `dmg` file produced for macOS should already contain the required file structure, and we should be able to handle custom URLs

sbt-native-packager should be enough to create those bundles.

## Comments:
This is login redirection ticket - possibly a ticket for @murbanczyk. (James Dunkerley - Feb 3, 2023)
<hr />

### PR DESCRIPTION
**Pull request by mwu-tow**
### Pull Request Description
This PR fixes #5239 by supporting the Windows-style of URL handling to support deep linking.

Windows spawns a new process for each URL, rather than sending a 'open-url' event to the existing process. Now the differences between the two platforms should be abstracted away.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

File: app/ide-desktop/lib/client/src/url-associations.ts
Comment (somebody1234) [inline app/ide-desktop/lib/client/src/url-associations.ts:23]: new file is missing sections - not sure it's required here, but afaict our code style requires sections in every file

File: app/ide-desktop/lib/client/src/project-management.ts
Comment (somebody1234) [inline app/ide-desktop/lib/client/src/project-management.ts]: should be avoidable by wrapping the `return` below in an `else {}`. alternatively, if the lint rule isn't useful, it can be removed completely

**Comment by somebody1234** [comment]
not explicitly approving since it'd be good to have another pair of eyes on this. also not sure this will work on linux (since it uses appimages) so i might try QAing this?

File: app/ide-desktop/lib/client/src/url-associations.ts
Comment (somebody1234) [inline app/ide-desktop/lib/client/src/url-associations.ts]: nit: `denote` -> `denotes`
the `denote` below is correct though

**Comment by somebody1234** [comment]
encountered a bug while testing so this took a while.
this seems to work on linux however:
- opens a second instance (correct) but that second instance opens a window which flashes before disappearing (bad!). we may want to consider *not* opening a window if electron detects that it is the second instance (if possible)
- i clicked "login with google". deep link was received, however ⚠️nothing happened⚠️. logs are below although likely doesn't explain the error
```
Got data from 'second-instance' event: '/tmp/.mount_enso-l26b4zj/enso,--allow-file-access-from-files,enso://auth?code=REDACTED&state=REDACTED'.
URL callback completed, releasing the lock.
Checking if 'enso://auth?code=REDACTED&state=REDACTED' denote a URL to open.
Parsed 'enso://auth?code=REDACTED&state=REDACTED' as URL: enso://auth?code=REDACTED&state=REDACTED. Protocol: enso:.
Got URL from second instance: 'enso://auth?code=REDACTED&state=REDACTEDf'.
Received 'open-url' event for 'enso://auth?code=REDACTED&state=REDACTED'.
```
- side note: may want to redact codes and/or states, manually with regex replacement, and/or by not printing out the link 6 separate times
  - alternatively, if the link is safe to post (e.g. time-sensitive and does not contain sensitive info and cannot be used to identify the user), then it might be good to add a log line explaining that

**Comment by somebody1234** [comment]
there were also two zombie `enso` processes, not sure if that's even related to this issue, or whether it's unavoidable linux weirdness

**Comment by mwu-tow** [comment]
@somebody1234 
Sorry for not making this clear before, the deep linking is not supported yet on Linux, it is expected to work on Windows and macOS only for now.

File: app/ide-desktop/lib/client/src/authentication.ts
Comment (wdanilo) [inline app/ide-desktop/lib/client/src/authentication.ts]: `${url.toString()}` should be quoted.

File: app/ide-desktop/lib/client/src/file-associations.ts
Comment (wdanilo) [inline app/ide-desktop/lib/client/src/file-associations.ts]: @somebody1234 why we still need such strange linter config in code? I remember we were talking about fixing linter config some time ago and making such annotations not needed anymore. What is the status of this topic? :)

File: app/ide-desktop/lib/client/src/url-associations.ts
Comment (wdanilo) [reply app/ide-desktop/lib/client/src/url-associations.ts:23]: Yes, its required, thanks for catching it!

Comment (wdanilo) [reply app/ide-desktop/lib/client/src/url-associations.ts:23]: Also, this is an exported function, it should have docs. Why linter did not catch it?

**Review by wdanilo — APPROVED** [review]

File: app/ide-desktop/lib/client/src/file-associations.ts
Comment (somebody1234) [reply app/ide-desktop/lib/client/src/file-associations.ts]: the cause of this specific one is bad eslint config, it should be able to be removed once #6267 is merged

File: app/ide-desktop/lib/client/src/url-associations.ts
Comment (somebody1234) [reply app/ide-desktop/lib/client/src/url-associations.ts:23]: i don't think the linter is currently set to error on missing docs, maybe i should look for an eslint plugin to try and enforce this?

*wait... it probably does have it built in. i'll look and see whether i accidentally disabled that rule

**Review by PabloBuchu — APPROVED** [review]

**Comment by PabloBuchu** [comment]
QA 🟢 

@mwu-tow I was able to make it finally running on my Windows.. We may think on some loader animation or at least disabling the button while user is waiting for the browser to open.. Other than that great work and thank you for making it running

**Review by somebody1234 — APPROVED** [review]

File: app/ide-desktop/lib/client/src/project-management.ts
Comment (somebody1234) [inline app/ide-desktop/lib/client/src/project-management.ts]: since the eslint fixes have been merged in, this *should* be redundant now

File: app/ide-desktop/lib/client/src/url-associations.ts
Comment (somebody1234) [inline app/ide-desktop/lib/client/src/url-associations.ts:167]: this should be redundant too

**Review by somebody1234 — APPROVED** [review]

**Comment by somebody1234** [comment]
i haven't `git checkout`ed this branch so i'm not 100% sure they can be removed - but it should be quick enough to check

File: app/ide-desktop/lib/client/src/project-management.ts
Comment (mwu-tow) [reply app/ide-desktop/lib/client/src/project-management.ts]: Done.

File: app/ide-desktop/lib/client/src/url-associations.ts
Comment (mwu-tow) [reply app/ide-desktop/lib/client/src/url-associations.ts:167]: This one seems to be still required.

Comment (somebody1234) [reply app/ide-desktop/lib/client/src/url-associations.ts:167]: ah, yeah, my bad - i think you can suppress it by explicitly doing `else { return }` though

**Pull request merged**
